processing/src/main/java/io/druid/segment/IndexMergerV9.java

```diff
@@ -0,0 +1,1321 @@
+import com.google.inject.Inject;
+import com.metamx.collections.bitmap.BitmapFactory;
+import com.metamx.collections.bitmap.ImmutableBitmap;
+import com.metamx.collections.bitmap.MutableBitmap;
+import com.metamx.collections.spatial.ImmutableRTree;
+import com.metamx.collections.spatial.RTree;
+import com.metamx.collections.spatial.split.LinearGutmanSplitStrategy;
+import com.metamx.common.IAE;
+import com.metamx.common.ISE;
+import com.metamx.common.guava.FunctionalIterable;
+import com.metamx.common.guava.MergeIterable;
+import com.metamx.common.io.smoosh.FileSmoosher;
+import com.metamx.common.io.smoosh.SmooshedWriter;
+import com.metamx.common.logger.Logger;
+import io.druid.collections.CombiningIterable;
+import io.druid.common.utils.JodaUtils;
+import io.druid.common.utils.SerializerUtils;
+import io.druid.query.aggregation.AggregatorFactory;
+import io.druid.segment.column.BitmapIndexSeeker;
+import io.druid.segment.column.Column;
+import io.druid.segment.column.ColumnCapabilities;
+import io.druid.segment.column.ColumnCapabilitiesImpl;
+import io.druid.segment.column.ColumnDescriptor;
+import io.druid.segment.column.ValueType;
+import io.druid.segment.data.ArrayIndexed;
+import io.druid.segment.data.BitmapSerdeFactory;
+import io.druid.segment.data.ByteBufferWriter;
+import io.druid.segment.data.CompressedObjectStrategy;
+import io.druid.segment.data.CompressedVSizeIndexedV3Writer;
+import io.druid.segment.data.CompressedVSizeIntsIndexedWriter;
+import io.druid.segment.data.GenericIndexed;
+import io.druid.segment.data.GenericIndexedWriter;
+import io.druid.segment.data.IOPeon;
+import io.druid.segment.data.Indexed;
+import io.druid.segment.data.IndexedIntsWriter;
+import io.druid.segment.data.IndexedIterable;
+import io.druid.segment.data.IndexedRTree;
+import io.druid.segment.data.TmpFileIOPeon;
+import io.druid.segment.data.VSizeIndexedIntsWriter;
+import io.druid.segment.data.VSizeIndexedWriter;
+import io.druid.segment.incremental.IncrementalIndex;
+import io.druid.segment.incremental.IncrementalIndexAdapter;
+import io.druid.segment.serde.ComplexColumnPartSerde;
+import io.druid.segment.serde.ComplexColumnSerializer;
+import io.druid.segment.serde.ComplexMetricSerde;
+import io.druid.segment.serde.ComplexMetrics;
+import io.druid.segment.serde.DictionaryEncodedColumnPartSerde;
+import io.druid.segment.serde.FloatGenericColumnPartSerde;
+import io.druid.segment.serde.LongGenericColumnPartSerde;
+import org.apache.commons.io.FileUtils;
+import org.joda.time.DateTime;
+import org.joda.time.Interval;
+
+import javax.annotation.Nullable;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.IntBuffer;
+import java.nio.MappedByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+
+public class IndexMergerV9
```

Comment:
I think this should replace the existing IndexMerger. Let's just have one version.